Various standalone migrations improvements

crisbeto opened the pull request:
Includes the following improvements to the standalone migration. There is more info in each of the commits:
1. Only excludes bootstrapped declarations from the first step of the migration, rather than all of them.
2. Migrates tests when switching the root module to standalone.
3. Doesn't copy over `ModuleWithProviders` when migrating tests.
4. Preserves the project's tsconfig when constructing the migration program.
5. Reduces the number of files that the migration has to load in order to speed it up.
6. Corrects any standalone declarations that might still be left in the `declarations` array from a previous migration. This is primarily targeted towards tests.

Fixes #48944.
Fixes #48971.

crisbeto committed:
fix(migrations): only exclude bootstrapped declarations from initial standalone migration

Currently the standalone migration is set up to skip any modules that have a `bootstrap` array with at least one element. This ends up being misleading for small apps who have everything in the root module.

These changes add some logic to only skip the root component.

Fixes #48944.

crisbeto committed:
fix(migrations): migrate tests when switching to standalone bootstrap API

The first step of the migration updates the tests of the component being migrated, however it also skips any bootstraped declarations. We get back to the these declarations when converting to the standalone bootstrapping APIs, but we weren't updating their tests.

These changes rework some of the logic so that we migrate any remaining tests as a part of the final step.

Relates to #48944.

crisbeto committed:
fix(migrations): don't add ModuleWithProviders to standalone test components

Fixes that we were copying all expressions in the `imports` array of the test to the `imports` of the component, including any potential `ModuleWithProviders`.

Fixes #48971.

crisbeto committed:
fix(migrations): preserve tsconfig in standalone migration

For the standalone migration we need to pass a couple of compiler flags which accidentally also overwrote the project's compiler options. These changes extend the options instead.

crisbeto committed:
fix(migrations): reduce number of files that need to be checked

Attempts to speed up the standalone migration by:
1. Not analyzing typechecker files.
2. Telling TS not to check libraries.

dylhunn commented on packages/core/schematics/ng-generate/standalone-migration/index.ts:
```diff
@@ -61,18 +61,19 @@ export default function(options: Options): Rule {
+      tree, tsconfigPath, basePath, undefined, undefined,
+      {
+        _enableTemplateTypeChecker: true,  // Required for the template type checker to work.
+        compileNonExportedClasses: true,   // Allows us to migrate some test classes.
```
What is actually meant by this?

crisbeto replied:
Reworded the comment.

dylhunn commented on packages/core/schematics/ng-generate/standalone-migration/to-standalone.ts:
```diff
@@ -351,41 +353,56 @@ function hasNgModuleMetadataElements(node: ts.Node): node is ts.PropertyAssignme
+  if (testBedImport || catalystImport) {
+    sourceFile.forEachChild(function walk(node) {
+      if (ts.isCallExpression(node) && node.arguments.length > 0 &&
+          ts.isObjectLiteralExpression(node.arguments[0])) {
```
Nit: What does `arguments[0]` signify here? Maybe consider a named variable to make it easier to follow

crisbeto replied:
Added a comment.

dylhunn reviewed (approved):
Looks good, and seems to fix the mentioned issue reports. (It's pretty involved though, so it's hard for me to be totally sure there are no new bugs.)

crisbeto commented on packages/core/schematics/ng-generate/standalone-migration/to-standalone.ts:
```diff
@@ -424,11 +424,53 @@ function findTemplateDependencies(
+
+  // If we can't analyze the `bootstrap` property, we can't safely migrate the module.
+  if (!ts.isPropertyAssignment(bootstrapProp) ||
+      !ts.isArrayLiteralExpression(bootstrapProp.initializer)) {
```
We're excluding anything that isn't an array literal, because we can't easily analyze it. E.g. `bootstrap: SOME_CONST`.

crisbeto committed:
fix(migrations): move standalone migrations into imports

Normally having a standalone declaration in the `imports` array is an error and something we handle in the conversion to standalone, but tests can end up in this situation, because apps may have separate tsconfigs for the main app and for tests.

These changes make it so that we move any incorrectly-defined standalone declarations, even if they aren't part of the current migration.

crisbeto commented:
I pushed one extra commit to handle any leftover invalid tests.

pkozlowski-opensource commented:
This PR was merged into the repository by commit c7926b57730c23f765a00d3dd9f92079c95e87e0.
